### ISSUE TITLE
Suggestions

### ISSUE THREAD

**Issue by leo60228** [v6cord/VVVVVV-CE#4]
stelpjo:
- ~~Non-repeating scripts without using flags~~ - done, hold down X when placing a script box to make it run only once
- ~~Fix cutscene bars appearing on one-time scripts with say or reply~~ - done, use `nobars()` to disable automatic bars
- ~~Fix looping playfile continuing after exiting level (playtesting or otherwise)~~ - done
- ~~Mode for flipping in mid-air/disabling flipping/jumping~~ - done, use `disableflip`, `enableflip`, `enableinfiniflip`, and `disableinfiniflip`
- Repo wiki? - not actionable
- ~~Command to go back to last checkpoint without dying~~ - done
- Script persisting (running a script behind another script) - not planned
- ~~Proper `stopscript` command~~ - done, use `stop()`
- ~~Labels and `goto` command~~ - ~~@AllyTally is working on this~~ @InfoTeddy did it instead - define a label with `$label$`, then go to it using something like `iftrinkets(0,$label$)`, if you're in another script do `iftrinkets(0,scriptname$label$)` instead
- Shape drawing (not really my idea) - @AllyTally is working on this
- ~~Variables~~ - ~~@AllyTally is working on this~~ - done, use `setvar(name [, contents])` to set a variable to the argument or the next line, and `addvar(name [, value])` to add/subtract an amount or concatenate a string, using the argument or the next line
- ~~Reload scripts in room without affecting entities~~ - done, use `reloadscriptboxes()`
- Sound remapping/disabling - not implemented
- ~~Fix delay when immediately loading a script when entering a room~~ - done

pixelator:
- ~~placeable upsidedown terminals~~ - done
- ~~`untilmusic()`, to wait until the music has fully faded in/out, but the only real usefulness that would have is covered by `nicefade()`~~ - done, also added `puntilmusic()`
- ~~`markmap(roomx,roomy,tiles1 tile id)`, kind of self explanatory~~ - done
- ~~`unmarkmap(roomx,roomy)`: removes all map markers from the specified room on the map~~ - done
- ~~inverse `stopmusic`. sometimes i want my track to start instantly yknow~~ - equivalent feature added, optional parameter of `play` sets fade in time in ms (just pass 0)
- ~~also an inverse `endtextfast`, maybe `speak_fast` and `speak_active_fast`~~ - done
- some `do()` modifier that will repeat for the duration that the player's hitbox is inside the script box. as a failsafe, maybe make it do nothing if the player has no control when they get to the loop command - not planned
- if `destroy(crewman,<color>)` is implemented, give it an extra argument for the crewman to be killed like the player (0 or 1) - not planned
- `creatergbcrewman` should be a thing. also should we use hex codes instead of three separate arguments for r, g, and b - unsure of the status on this

**Comment by Fussmatte** [comment]
What I meant by repo wiki was a wiki on the repo containing explanation and documentation for all the new features.

**Comment by Fussmatte** [comment]
Suggestions:

- ~~Fix `mapimage()` (and custom maps in general) not working with `gamemode(teleporter)`.~~ - done
- ~~Implement a command that shows all the coordinates on the map permanently.~~ - done
- ~~Implement a command that shows all the coordinates on the map temporarily (perhaps until the script ends.)~~ - done
- ~~Fix the memory leakage from using `mapimage()` multiple times.~~ - done
- ~~Implement a command that loops `mapimage()` in succession by loading images incrementally, eg `mapimage(graphics/filename$.png,000,134,5)`, with $ being replaced with 000-134 automatically, and putting a 5-frame delay between each usage of mapimage. Maybe this is too crazy, but it'd be cool for animations using `gamemode(teleporter)`.~~ - use variables

**Comment by Fussmatte** [comment]
In hindsight, just implementing variables that can be used combined with filenames in CE commands would do the same job as the last one.

**Comment by leo60228** [comment]
> Fix mapimage() (and custom maps in general) not working with gamemode(teleporter).

> Implement a command that shows all the coordinates on the map permanently.

> Fix the memory leakage from using mapimage() multiple times.

Done. (new command is `disablefog`)

> Implement a command that shows all the coordinates on the map temporarily (perhaps until the script ends.)

Does using `disablefog` and `enablefog` work?

> In hindsight, just implementing variables that can be used combined with filenames in CE commands would do the same job as the last one.

@AllyTally Are you planning this?

> Implement a command that loops mapimage() in succession by loading images incrementally, eg mapimage(graphics/filename$.png,000,134,5), with $ being replaced with 000-134 automatically, and putting a 5-frame delay between each usage of mapimage. Maybe this is too crazy, but it'd be cool for animations using gamemode(teleporter).

I'd rather just use variables if possible.

**Comment by leo60228** [comment]
Wait, I haven't pushed yet.

**Comment by leo60228** [comment]
Done.

**Comment by Fussmatte** [comment]
> markmap(roomx,roomy,tiles1 tile id), kind of self explanatory - it isn't imo

What I think pixelator meant by this was that you should be able to make a map marker (like the T's representing teleporters or S's representing trinkets in the main game) using any tile from tiles1.

Thanks for the fixes, btw!

**Comment by AllyTally** [comment]
> > In hindsight, just implementing variables that can be used combined with filenames in CE commands would do the same job as the last one.
> 
> @AllyTally Are you planning this?

If you set the variable `this` to `green`, then something like `changeplayercolour(%this%)` would turn the player green. It's basically just a simple find/replace, so it'll most likely work with commands that need filenames. If not, that's a huge bug!

**Comment by Fussmatte** [comment]
More suggestions:

- ~~Enemy speed~~ - done
- ~~Gray Warp Zone tileset in editor~~ - done
- ~~Fix green enemies in gray Warp Zone tileset~~ - done

**Comment by InfoTeddy** [comment]
> Fix green enemies in gray Warp Zone tileset

As of c47061d9d8c384ed287b9689178b56fe2bf63035, this is fixed.

**Comment by InfoTeddy** [comment]
> placeable upsidedown terminals

As of ca9867e09ce45a8dbb39cd0457a1ca7f244b10f9 and 02c267eeae0a466cff3e571b67be4a21f7ea7c6c, this is added.

**Comment by InfoTeddy** [comment]
> Fix cutscene bars appearing on one-time scripts with say or reply

I am interested in figuring out a good theoretical solution to this. I don't have a good one yet.

**Comment by leo60228** [comment]
I was thinking to insert the cutscene bars before the first say/reply and remove them after the last one.

**Comment by leo60228** [comment]
btw @InfoTeddy, if you have owner perms in an organization you can edit other people's comments

**Comment by Fussmatte** [comment]
More suggestions:
- Whenever ``onkeypress()`` is implemented, maybe add a second scriptable action button mapped to the X key (and the X button on an Xbox controller - that is, the left action button).
- ~~When variables are implemented, add some constants for the player's X and Y positions.~~ - done, `%player_x%` and `%player_y%`
~~- Fade out the "[Press ENTER to return to editor]" text after a few seconds~~
- ~~Add a way to do ``drawtext()`` with a border (i.e. with ``dwgfx.bprint()``)~~ - done, 7th argument of `drawtext()` is the border, type "true" to draw border

**Comment by InfoTeddy** [comment]
@leo60228 Oh, okay.

**Comment by InfoTeddy** [comment]
> Fix cutscene bars appearing on one-time scripts with say or reply

As of a3cd8f765f1c7a59f347fcc09b106e0aeec22ede, you can use `nobars()` to disable automatic cutscene bar placement.

**Comment by InfoTeddy** [comment]
> Reload scripts in room without affecting entities

As of a62adec85549e95e9ae7c8e60e4fdf80f8bde78d, you can use `reloadscriptboxes()`.

**Comment by InfoTeddy** [comment]
> Gray Warp Zone tileset in editor

Added in b989857b9cb6b96a275ab8df4b37554beca1da06.

**Comment by InfoTeddy** [comment]
> Mode for flipping in mid-air/disabling flipping/jumping 

Both added in a332ed36cbbdeb232db64bfbbfa8cc87afebf1fb and 05f8e715928b33aa4d3578cafe9f0098bc80543a for toggling flipping, and d2479b357583ad646c98dfea8d5b9a61de313192 and e14ac342bb2952f4b81aef0ccc8a20bb8d08be86 for toggling mid-air flipping. Use `disableflip` and `enableflip` for toggling flipping and `enableinfiniflip` and `disableinfiniflip` for toggling mid-air flipping.

**Comment by InfoTeddy** [comment]
> untilmusic(), to wait until the music has fully faded in/out, but the only real usefulness that would have is covered by nicefade()

Whatever this `nicefade()` is supposed to be, I can't find it anywhere and it's not in the Discord. So I can't tell what Pixelator wants a `nicefade()` to do.

**Comment by InfoTeddy** [comment]
> Proper `stopscript` command

Do we need this? `loadscript(stop)` works fine.

**Comment by InfoTeddy** [comment]
> `untilmusic()`, to wait until the music has fully faded in/out, but the only real usefulness that would have is covered by `nicefade()`

This is useful apart from a not-yet-added `nicefade()` because a while ago a second optional argument to `play()` was added, being the fade-in time for the track in milliseconds.

And as of 2f8ef9213c223c093a2c02c64ed8d27132d87fff, `untilmusic()` has been added, along with `puntilmusic()`.

**Comment by leo60228** [comment]
I'm guessing it was a typo of niceplay.

**Comment by InfoTeddy** [comment]
> Labels and `goto` command

As of 66cbae3bb1765a144e755af39b99e6c45f4b309e, this is added. Define a label using `$label$`, then to jump to it, use it as a script name in a command like `iftrinkets`, e.g. `iftrinkets(0,$label$)`. If you're in another script, simply use `scriptname$label$` as a script name instead.

**Comment by InfoTeddy** [comment]
> `creatergbcrewman` should be a thing. also should we use hex codes instead of three separate arguments for r, g, and b - unsure of the status on this

Why should we use hex codes instead?

**Comment by InfoTeddy** [comment]
> Fix delay when immediately loading a script when entering a room

As of a2fda687b806893f3cb8942aa9dc7da3198f7f66, this is fixed.

I personally checked with libTAS that the script executes immediately upon entering a room, btw.

**Comment by InfoTeddy** [comment]
> Enemy speed

Please elaborate on this one, @Stelpjo. Do you mean you want a room property to control the room's edentity enemy speed just like the platv room property?

**Comment by Fussmatte** [comment]
> > Enemy speed
> 
> Please elaborate on this one, @Stelpjo. Do you mean you want a room property to control the room's edentity enemy speed just like the platv room property?

Yes, that would be nice!

**Comment by leo60228** [comment]
Discord bot to automatically upload levels - bot at https://github.com/leo60228/levelbot, client not implemented

**Comment by InfoTeddy** [comment]
> also an inverse `endtextfast`, maybe `speak_fast` and `speak_active_fast`

As of 54925bb2f6e175ff75fb300ef76aee893e9a8d1e, these have been added.

**Comment by InfoTeddy** [comment]
Variables have been added (they've already mostly been added, I just added support for saving and loading them from a quicksave).

Use `setvar(name [, contents])` to set a variable to either the given argument or to what's on the next line. Use `addvar(name [, value])` to add or subtract an amount from a number, or to concatenate a string to a variable, using either the given argument or whatever's on the next script line.

**Comment by Fussmatte** [comment]
More suggestions:
- ~~Add ``ifvar(name,less/lesseq/greater/greatereq,value,script)``~~ - ~~@AllyTally is working on this~~ - done
- Allow loading additional font data from ``font2.png`` and ``font2.txt`` (possibly 3, 4, etc as well) - this would be helpful for levels to be able to add additional custom characters while still allowing ``font.png`` and ``font.txt`` to be appended to in updates

Besides that and fixing automatic custom asset loading, I think a lot's already been good enough for me.

**Comment by Trinket9** [comment]
Suggestions:
- ~~F11 to change enemy speed in the editor - ideally you'll get a window with a number to input~~ - done, use ~~the `[` and `]` keys~~ Ctrl+`,`/`.` (mnemonic: Ctrl+`<`/`>`) when placing them down
- F12 to change platform speed in the editor - this actually exists in the main game, but a way to do it without modifying the file through notepad would be nice. (or just do it like enemies)
- Center the direction signs on non 16x16 enemies
- ~~Extend the 20x20 map limit to 30x30 or 25x25~~ - done, map limit is now 100x100
- ~~Teleporters, ideally ones that work, alternatively as a placeable entity that lights up on touch, acts as a checkpoint and executes a script on pressing enter~~ - all done, except for the teleporter script (just set teleporter script with vanilla command `teleportscript(custom_<script>)`
- Show pixel coords (multiples of 8, like in ved) while hovering over a tile (while not in tower mode) - to the right of the room coords
- _maybe_ a way to set a diagonal enemy/platform direction movement type
- ~~`ifinvincible(script)` - checks if the player has enabled invincibility or not, if yes go to script~~ - not being implemented
- ~~make `warpdir(x,y,0)` switch to the correct background (lab, tower, WZ)~~ - done, but Lab and Warp Zone were already done at time of suggestion
- ~~make editing an entity's position through notepad a working way to set entities at any pixel position (so that the game will not automatically correct them)~~ - done
- ~~make SS1 -1 tileset available through the editor and make its entities gray (like in gray WZ)~~ - both done
- a variant of warp token that either preserves the player's gravity or sets their gravity to upside down
- ~~remove the `settile` thing from all rooms where it happens but exactly (12,8) 1-indexed which is prize for the reckless~~ - upstreamed, it has been removed entirely from custom levels
- ~~working enemies, platforms, quicksand, warptokens etc in towers~~ - done
- ~~a correctly working one-way tile that points up (doesn't freeze your flip while standing on it) - in a different spot on `tiles.png`/`tiles2.png` to avoid breaking any levels~~ - done, however no new one-ways have been added, instead the original ones have been fixed in VCE levels only
- fix enemy and platform collisions with one-way tiles

**Comment by InfoTeddy** [comment]
> remove the settile thing from all rooms where it happens but exactly (12,8) 1-indexed which is prize for the reckless

If we're fixing this (which I don't really plan on doing anyway), I'd rather we just make it so it only happens in the main game. I don't think we need this hardcoded behavior anywhere in custom levels.

**Comment by InfoTeddy** [comment]
> Extend the 20x20 map limit to 30x30 or 25x25

As of b1bfe2e8a7c1270e7234a785189e3c4e5d173e4d, this limit has been raised to 100x100.

**Comment by PIXELATOR37** [comment]
why hello, it is I, the person who made an account to comment on this

> > `creatergbcrewman` should be a thing. also should we use hex codes instead of three separate arguments for r, g, and b - unsure of the status on this
> 
> Why should we use hex codes instead?

i just didn't want to raise the number of arguments the command has to around 10(!). seeing as it would be more consistent to use three separate arguments instead, that should be what's done, but that will make it 100% impossible to refer to the crewmate in other commands

anyway, i'm putting some more suggestions:
* left/right towers (very far fetched but had to put it here) - not planned
* fitting more enemy movement types for the editor (like the ones used by zzt centipedes and lies)
* adding the vertigo and box (unused) enemy types to the editor
* `prompt(left/right/action)`. it's basically a prompt to continue the script without having to use a text box - planned, but only for action
* based on the above, `p_prompt(input,script,frames)`, which will go to the specified script if you perform the input in the specified number of frames, but continue the script if time runs out  - onkey() is planned for when async scripts happen
* a system for adding custom colors. you guys can figure out the rest - not planned
* ~~usable orange text boxes without `textrgb`. cmon i know the color is there~~ - done
* `setspeed` and `setvelocity` for the Y direction - unsure of status
* combining all the commands for flipping, finalstretch, etc. into single commands with a simple argument for `on` or `off` - unsure of status
* ~~screen flashes in any rgb color~~ -  `drawrect()`

**Comment by AllyTally** [comment]
> More suggestions:
> 
> * Enemy speed

As of 022b2f50123158a2736370494aa1344e545464c7, this has been implemented. Use `[` and `]` to change the speed of enemies you're about to place in the editor.

**Comment by mothbeanie** [comment]
some incredibly specific suggestions of my own!

* `typewritertext(on,speed)`, alternatively `typewritertext(speed)`: for textboxes. instead of the whole content of the textbox appearing at once, each character would appear one by one, until all characters are visible. maybe would show a `▼` afterward that blinks to indicate that its done. `speed` would be the frames to wait between each character.
* `typewritertext(off)`: turns the above off. alternatively, could just be `typewritertext(0)`
* `starspeed(speed,direction)`, alternatively `starspeed(speed,directionx,directiony)` sets the speed of the scrolling star background. `speed` would be the speed (with 0 as static). `direction` would be 0 (right to left), 1 (down to up), 2 (left to right), 3 (up to down). alternatively, directionx and directiony could be how fast it moves in each direction, for diagonal scrolling.
* `setbackground(image,speed,direction)`, alternatively `setbackground(image,speed,directionx,directiony)`: sets the background of the room (of all rooms until you use it again???) `image` would be the image file. `speed`, `direction`, `directionx`, and `directiony` would have the same behavior as they would for starspeed.

**Comment by mothbeanie** [comment]
(again, these are incredibly specific and mildly implausible so if none of them are possible/are extremely low priority thats OK)

**Comment by InfoTeddy** [comment]
`starspeed()` would have to be a *speed multiplier* instead, because the stars move at varying (random) speeds.

**Comment by PIXELATOR37** [comment]
huh, it always seemed to me that there were two "groups" of stars, one with a set faster speed and the other with a set slower speed.
though, that gives me an idea for a command -
* `starcol(r,g,b,r2,g2,b2,r3,etc...)` sets the rgb color of the background stars. if more than one color is entered, it could be randomized between all of them. for the further bg stars, either you could make them be a darker multiple of the normal ones, or make a separate command for it
* if that's possible, a bg color changing command could also be cool

**Comment by InfoTeddy** [comment]
> huh, it always seemed to me that there were two "groups" of stars, one with a set faster speed and the other with a set slower speed.

There is, in a sense, two "groups". One of them is light-colored, the other is dark-colored. If a star's speed is less than 7 pixels per frame, it becomes the dark-colored speed. But within these two groups, they all have varying random speeds.

**Comment by AllyTally** [comment]
> But within these two groups, they all have varying random speeds.

Due to the star speed being an `int`, there's really only three speeds

**Comment by InfoTeddy** [comment]
> usable orange text boxes without `textrgb`. cmon i know the color is there

Not sure if you just misremembered, but the command to create a text box with RGB values is `textcolor` or `textcolour`, not `textrgb`.

**Comment by InfoTeddy** [comment]
> usable orange text boxes without `textrgb`. cmon i know the color is there

As of 6162d4bf983dd7f69dfb1239fa14070d593d53b1, this has been added.

**Comment by InfoTeddy** [comment]
> i just didn't want to raise the number of arguments the command has to around 10(!). seeing as it would be more consistent to use three separate arguments instead, that should be what's done, but that will make it 100% impossible to refer to the crewmate in other commands

It's really cumbersome to have to refer to crewmates by saying (115,200,5) all the time, no matter if it's collapsed into one argument by saying 0x73C805 instead. Plus, there's still the problem of being unable to neatly refer to multiple crewmates with the same color.

A better solution would just be being able to name crewmates instead.

**Comment by InfoTeddy** [comment]
> Show pixel coords (multiples of 8, like in ved) while hovering over a tile (while not in tower mode) - to the right of the room coords

The right space of the room coordinates is already reserved for denoting the current alt state if you are not in the main state and denoting the current tower if you are in one. We can't add pixel coordinates to the right of the room coordinates or they'd most certainly go offscreen.

**Comment by leo60228** [comment]
@InfoTeddy "A better solution would just be being able to name crewmates instead." this is already implemented lmao

**Comment by InfoTeddy** [comment]
Then the remaining problem with such a `creatergbcrewman` is the fact that the `colour` attribute of entities always goes through the `Graphics::setcol()` system.

**Comment by InfoTeddy** [comment]
> make SS1 -1 tileset available through the editor and make its entities gray (like in gray WZ)

As of 12340a2f888bb7aff4fee9b9ab268ff12c311cc1 and 9155f9baaf14abc946882147e904cbaa68a15b4f, these have been added.

**Comment by InfoTeddy** [comment]
> make warpdir(x,y,0) switch to the correct background (lab, tower, WZ)

You're a bit late to the party - I've already [gotten it fixed in the official VVVVVV repository](/TerryCavanagh/VVVVVV/pull/86). So two-thirds of this suggestion is really, really redundant.

That said, the Tower background is something we added, so we'll have to fix that...

**Comment by InfoTeddy** [comment]
Fixed `warpdir` setting current room warp background to none in a Tower room in d990c4982ee0b8f5f56750db36ad46068829ce9f.

**Comment by InfoTeddy** [comment]
> make editing an entity's position through notepad a working way to set entities at any pixel position (so that the game will not automatically correct them)

As of bded94a19339ed2cbe1ba947b2400fd2f60f5a1b, this is added. You can edit the `subx` and `suby` attributes of entities in Notepad.

**Comment by InfoTeddy** [comment]
> Proper `stopscript` command

Stelpjo, you added this (named `stop`) but forgot to strikethrough your own suggestion.

**Comment by InfoTeddy** [comment]
> When variables are implemented, add some constants for the player's X and Y positions.

This was added (`%player_x%`, `%player_y%`) but the suggestion didn't get strikethroughed at the time of implementing it.

**Comment by InfoTeddy** [comment]
> Add `ifvar(name,less/lesseq/greater/greatereq,value,script)`

This was added, but the suggestion didn't get strikethroughed when it was added.

**Comment by InfoTeddy** [comment]
> working enemies, platforms, quicksand, warptokens etc in towers

Enemies already work just fine (although their color doesn't change, which is minor but will be fixed). Platforms and quicksand render weirdly. Platforms don't move, but quicksand is already functional. Warp tokens are the most broken here, so half of the things you listed are already more-or-less functional.

**Comment by InfoTeddy** [comment]
> Add a way to do `drawtext()` with a border (i.e. with `dwgfx.bprint()`)

Stelpjo, just like with `stop`, you added this but forgot to strikethrough your own suggestion.

**Comment by Stelpjo** [comment]
> > Proper `stopscript` command
> 
> Stelpjo, you added this (named `stop`) but forgot to strikethrough your own suggestion.

oh

> > Add a way to do `drawtext()` with a border (i.e. with `dwgfx.bprint()`)
> 
> Stelpjo, just like with `stop`, you added this but forgot to strikethrough your own suggestion.

_oh_

**Comment by Fussmatte** [comment]
Also, can someone fix sounds not loading with the `reloadresources()` function? I tried to, but i havent got a clue what im doing

**Comment by leo60228** [comment]
test

**Comment by AllyTally** [comment]
> test

Playtesting has existed since f7c0321b715ceed8e87eba2ca507ad2dc28a428d.

**Comment by InfoTeddy** [comment]
> combining all the commands for flipping, finalstretch, etc. into single commands with a simple argument for `on` or `off`

Do we *really* need to do this?

Also, it'd be inconsistent with the vanilla commands. And I'm not changing the vanilla commands because that will break levels.

**Comment by Trinket9** [comment]
More suggestions:
- a LIES and cloud enemy that behave like normal enemies rather than being fixed to a specific movement type and speed
- an option to set edentity conveyor speed (like enemies), and make them actually animate as fast as their speed rather than always the same. that would be pretty cool
- Ctrl+Z to undo, Ctrl+Y to redo
- enemy movement type which causes enemies to appear and disappear repeatedly at an interval determined by speed
- an option to set bounds for each individual enemy/platform
- ~~fix enemy movement type 1 bugging when attempting to warp through the top of a screen~~ - fixed
- Add 4 extra arguments to createentity to set enemy or platform bounds for the created entity (there isn't really a way to do it with less arguments unfortunately, you need 2 for x and 2 for y)
- Add an option to customize keybinds
- Correctly number the music in the editor and add Path Complete (use 0-15 instead of 1-11 and N/A)
- Support other entity movement types in the editor
- Make it possible to place entities at subtile positions in the editor - maybe while holding Shift or Ctrl while on an entity tool

**Comment by leo60228** [comment]
The numbering in the editor is the same as the `music` simplified command.

**Comment by InfoTeddy** [comment]
> `play` simplified command.

`play` is an internal command.

**Comment by InfoTeddy** [comment]
> Non-repeating scripts without using flags

As of 4ce23bfb217d91c3cc53ecdc45029c60af2ee86b and a3346ca420a2df34679da2b57a5a52026ea8c11b, you can hold down X when placing a script box to make it run only once.

**Comment by InfoTeddy** [comment]
> fix enemy movement type 1 bugging when attempting to warp through the top of a screen

This isn't a bug specific to any enemy or any movement type - it looks like the bounds of enemies and platforms are just bugged in general in warping rooms.

**Comment by InfoTeddy** [comment]
> fix enemy movement type 1 bugging when attempting to warp through the top of a screen

Fixed in ef023648767b036e18ead401fc40146de009853a.

**Comment by InfoTeddy** [comment]
>> remove the settile thing from all rooms where it happens but exactly (12,8) 1-indexed which is prize for the reckless

> If we're fixing this (which I don't really plan on doing anyway), I'd rather we just make it so it only happens in the main game. I don't think we need this hardcoded behavior anywhere in custom levels.

Furthermore, this should be fixed upstream, and not only in VCE. In fact, there are a *lot* of things that should be sent upstream instead of being VCE-only (such as the optimized levels list loading, although it seems to be a bit buggy right now, so it's probably best that it hasn't been sent upstream yet...).

**Comment by InfoTeddy** [comment]
I've sent a fix upstream for the `map.settile(18, 9, 59);` bug: TerryCavanagh/VVVVVV#136

**Comment by InfoTeddy** [comment]
Alright, the `map.settile(18, 9, 59);` fix has been upstreamed.

**Comment by InfoTeddy** [comment]
> Fade out the "[Press ENTER to return to editor]" text after a few seconds

Not sure why this should be VCE-only. If this is going to be added, it should be upstreamed as well.

**Comment by InfoTeddy** [comment]
>> combining all the commands for flipping, finalstretch, etc. into single commands with a simple argument for `on` or `off`
>
> Do we *really* need to do this?
>
> Also, it'd be inconsistent with the vanilla commands. And I'm not changing the vanilla commands because that will break levels.

Actually, I guess we could do this for not only VCE commands but vanilla commands as well, and keep the old ones around for backwards compatibility. But still, I ask if we actually need to do this or not.

**Comment by InfoTeddy** [comment]
> working enemies, platforms, quicksand, warptokens etc in towers

This suggestion is already more-or-less finished, now the entities that don't work are script boxes and terminals. Enemies were already working, but now they change color with the tower. Platforms and quicksand now render properly and change color with the tower. Platforms used to not move their collision. Quicksand used to not respawn upon death. And warp tokens now work properly.

**Comment by AllyTally** [comment]
Placing down terminals that don't create activity zones would be nice. Adding it to the UI would be simple--holding down Z while placing a terminal.

**Comment by AllyTally** [comment]
`moveplayersafe(x,y)` - `moveplayer(x,y)` but it doesn't move the player inside walls. If the destination will be inside of a wall, push them out of it, so they'd be as close as possible.
`ignoredamage(bool)` - If this is true, then the player should be able to walk through spikes and enemies without dying.

**Comment by AllyTally** [comment]
~~`csay([lines[, color]])` - `say([lines[, color]])` except it uses `position` instead of `customposition`.~~ - done

**Comment by InfoTeddy** [comment]
> `csay([lines[, color]])` - `say([lines[, color]])` except it uses `position` instead of `customposition`.

As of cb53b9d7e861c2d7ea3dfe3dbf5d2a7d491148f2, this is added. Also, `csayquiet()` is also added in 84fc46a0c8c1fba71889d1867c00adc3390afa34.

**Comment by AllyTally** [comment]
I feel like there should be a way to completely pause the players movement (without affecting anything else). `freezeplayer`/`unfreezeplayer` could work, or `playerfreeze(on/off)`
When the player is frozen they shouldn't be able to move or fall (but they should still interact with stuff, like being able to die, being flipped from gravity lines, collecting trinkets etc). All it'd really do is pause the players movement. This would be extremely useful for things like powerups, or custom abilities which mess with the players movement using a frozen player and `moveplayersafe`.

**Comment by AllyTally** [comment]
`char_at(var,string,index)`
basically just saves words[2][ss_toi(words[3)] to `var`
words[3] should be clamped

**Comment by AllyTally** [comment]
~~`disablepause(on/off)` - make enter do nothing, this would also affect playtesting (as people can just use escape)
this would be useful for ifkey stuff~~ done

**Comment by InfoTeddy** [comment]
> `disablepause(on/off)`

I'd rather name it `togglepause(on/off)` so the negative doesn't confuse anyone.

**Comment by AllyTally** [comment]
that makes more sense, originally i was thinking itd be `pause(on/off)` but `pause()` could be its own command that pauses the game, or at least people would think itd do that

**Comment by AllyTally** [comment]
this could be pretty hard:
`%script_x%`, `%script_y%`, `%script_w%`, `%script_h%` - position and size of the script box or activity zone that called the script
`%script_type%` - `box` (script box), `terminal`, `activityzone` or `script` (another script using load or smth)
if the script was loaded from another script, these variables should be kept from the last script that called it
if the script is called from a terminal, they should be the 2x2 area that the terminal takes up (or 2x3, depending if terminals are *actually* 2x3 or not)

**Comment by InfoTeddy** [comment]
> (or 2x3, depending if terminals are *actually* 2x3 or not)

Only edentity terminals are 2x3 (if unflipped, in VCE flipped edentity terminals are 2x2), but the terminal that gets `createentity`ed for each edentity terminal is 2x2.

**Comment by InfoTeddy** [comment]
> ~a correctly working one-way tile that points up (doesn't freeze your flip while standing on it) - in a different spot on `tiles.png`/`tiles2.png` to avoid breaking any levels~

This has been added, however we have not added another set of one-ways. Instead, we fixed the original set, but only in VCE levels.

**Comment by mse-k** [comment]
> > untilmusic(), to wait until the music has fully faded in/out, but the only real usefulness that would have is covered by nicefade()
> 
> Whatever this `nicefade()` is supposed to be, I can't find it anywhere and it's not in the Discord. So I can't tell what Pixelator wants a `nicefade()` to do.

`nicefade(x)` changes the music to x smoothly (?) i dont really know that it uses to fade "nicely" but it isnt just instant change, so he want it to wait until it actually switches tracks

**Comment by AllyTally** [comment]
> `nicefade(x)` changes the music to x smoothly (?) i dont really know that it uses to fade "nicely" but it isnt just instant change, so he want it to wait until it actually switches tracks

The correct command is `niceplay(x)`, and they can use a simple `delay(x)` command.

**Comment by InfoTeddy** [comment]
> `nicefade(x)` changes the music to x smoothly (?) i dont really know that it uses to fade "nicely" but it isnt just instant change, so he want it to wait until it actually switches tracks

`nicefade` was just a typo on the user's part, they meant `niceplay()`.

**Comment by mse-kez** [comment]
> ~~Variables - @ AllyTally is working on this~~ - done, use setvar(name [, contents]) to set a variable to the argument or the next line, and addvar(name [, value]) to add/subtract an amount or concatenate a string, using the argument or the next line

is there a way to run a script if a variable is x (somthing like say `ifvar(name [, value])` where if value is omitted it will check if it existsand will otherwise see if it is equal to it [there could even be a third param to set what i compairs say if its less than ect])

**Comment by InfoTeddy** [comment]
Yes, `ifvar()` exists. Although I don't think it's intended to be used to check if a variable exists.

**Comment by mse-k** [comment]
ahh ok thx i was suggesting what it would/could do if it want a thing yet

**Comment by mse-k** [comment]
can i have the syntax for it

**Comment by InfoTeddy** [comment]
It's documented in [FEATURES.md](/v6cord/VVVVVV-CE/blob/master/FEATURES.md) (Ctrl+F for `ifvar()`):

> `ifvar(var,operator[,value],script)` - go to _script_ if a variable equals/isn't equal to/is greater than/is lesser than/is greater or equal to/is lesser or equal to _value_ - if _value_ not given, uses the next line

**Comment by ncrecc** [comment]
Advanced autotiling that accounts for the thin walls and extra corners in Lab tileset and the single tiles in most tilesets (e.g. tiles2 tile 225), latter isn't as important

**Comment by AllyTally** [comment]
> `disablepause(on/off)` - make enter do nothing, this would also affect playtesting (as people can just use escape)
> this would be useful for ifkey stuff

as of https://github.com/v6cord/VVVVVV-CE/commit/094c229c0c479a92f133e306f89c3d1880f7370b, this now exists as `togglepause(on/off)`

**Comment by AllyTally** [comment]
`%entities%` - ~~The number of entities in the room, useful for looping through all entities in the room to do something for each of them~~ done in 588e0eeefb7a9adb7d26ec7b2c7c1f4dd794f173
`register(commandname,script)` - Run scripts like commands! Could possibly have a namespace argument so you would have to invoke a custom command like `std:cool_command(a,b)`. The variables `arg1`, `arg2` etc would be set to the arguments you pass in the command, and `args` would be the number of arguments passed. If arguments aren't passed, they should be set to an empty string

**Comment by AllyTally** [comment]
%cursor_x%, %cursor_y% - cursor position, clamped between 0,0 and 320,240

**Comment by PIXELATOR37** [comment]
- ~~%altstate% - built-in variable for the altstate the player is in~~ done

- `createcollision(x,y,w,h)` - create a region that acts like a solid wall for all entities

- `imagedata(id,x,y[,filename[,centered[,alpha[,background[,blend]]]]])` - manipulate an existing image that's been created with `drawimagepersist`; changing the filename of a loaded image seems hacky, but it could allow for easier image animation, as the current alternative is having to remove and redraw the image for every new sprite

- two new optional arguments to `getvar`: n, which is the number of characters to read from _var2_, and `start` or `end`, which determines whether to read said characters forward from the start of the variable, or back from the end, respectively

**Comment by iTom33ro** [comment]
Hey, I have few suggestion:

- To make `R` key work in `ifkey()`

- To keep raw color in `createentity()`

- To add a new command `customenemy(x,y,mt,ms,bx,by,bw,bh,c,t,hw,hh,f,fs,h)`, which:

    -  `x` and `y` are enemy position;

    -  `mt` and `ms` are movement type and movement speed;

    -  `bx`, `by`, `bw` and `bh` are bound position and bound size;

    -  `c` is enemy raw color;

    -  `t` is enemy tile (from sprites.png);

    -  `hw` and `hh` are hitbox size;

    -  `f` and `fs` are frames and frame speed;

    -  `h` is if enemy hurt or not.

**Comment by AllyTally** [comment]
`clearimagecache()` - simple command that just empties the image cache that gets filled using drawimage/loadimage. this is     like   super duper important for levels that have like, a drawimage every room
